src/thunderbots/software/test/ai/evaluation/calc_best_shot.cpp

```diff
@@ -0,0 +1,166 @@
+/**
+ * This file contains the unit tests for evaluation functions
+ * in calc_best_shot.cpp
+ */
+
+#include "ai/hl/stp/evaluation/calc_best_shot.h"
+
+#include <gtest/gtest.h>
+
+#include "geom/util.h"
+#include "shared/constants.h"
+
+TEST(CalcBestShotTest, calc_best_shot_with_vector_of_points_test)
+{
+    Field f                      = Field(10, 8, 10, 8, 3, 1, 1);
+    std::vector<Point> obstacles = {Point(2, -3), Point(6, 2), Point(1, 4)};
+    const Point p                = Point(0, 0);
+    double radius                = 1.0;
+    const Point p1               = Point(f.length() / 2.0, -f.goalWidth() / 2.0);
+    const Point p2               = Point(f.length() / 2.0, f.goalWidth() / 2.0);
+    EXPECT_EQ(angleSweepCircles(p, p1, p2, obstacles, radius * ROBOT_MAX_RADIUS_METERS),
```

Comment:
My gut is telling me these tests should compare against actual "hand calculated" values rather than the result of `angleSweepCircles` because this is too close to just repeating the actual implementation of the function.

@garethellis0 ?

Reply:
Yeah, I agree. I'd probably just do it on paper and leave a note indicating how it was calculated.

Reply:
@HaomiaoZ you can modify the test case so that the hand calculations are easier though. So you get nice fractions rather than long decimals. Easier to do by hand and less confusing to write out.

Reply:
Changed. Currently the hand calculation values match the function output, but both `calcBestShotOnEnemyGoalAll()` functions will generate a extra `std::pair` with the location of `f.enemyGoalpostPos()`. Looking into that right now
![Screenshot from 2019-03-17 13-41-14](https://user-images.githubusercontent.com/43716969/54497686-fac76800-48ba-11e9-813e-03969df93fa9.png)
![Screenshot from 2019-03-17 13-47-21](https://user-images.githubusercontent.com/43716969/54497709-36fac880-48bb-11e9-97c8-10f7fc0ae4a1.png)

Reply:
![webwxgetmsgimg_rotated](https://user-images.githubusercontent.com/43716969/54498006-b047ea80-48be-11e9-8e70-64a24cf34587.jpeg)
This is the layout for hand calculation. Two shaded circles are the obstacles. 0.15, 1.12, 1.13 and 0.5, 1.2, 1.3 are triplets that can form right angled triangles. Locations and radii are chosen so that when tangent, the triangles form by radius, tangent line and line between shooting points and the center of circle is known.

Reply:
Looks reasonable :+1:

Reply:
![IMG_4288_cropped](https://user-images.githubusercontent.com/43716969/54862006-bee43680-4cf0-11e9-842a-cf1d2521a855.jpg)

@MathewMacDougall I went over the implementation of `angleSweepCirclesAll()` and find out that as long as the obstacle blocks `posPost`, then there will be a element in the return vector with direction along `posPost` and 0 angle. If the situation is similar to the layout on the **right** side, then there will be two entries with vector along line `one` and `two` and 0 angle.  So far to me it seems that `angleSweepCirclesAll()` only produces extra elements. Other elements seem to be correct. What would be your suggestion on this problem?

Reply:
Oh that's a weird bug. This is definitely something we'll want to fix, likely as part of this PR.

If you want to get started on this you can start debugging the `angleSweep...` functions to try figure out why this happens. Making some very simple test cases separately for `angleSweepCircles` would be a good idea.

We can talk at the meeting and go over this in more detail together.